internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package variables
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	structuraldefaulting "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ValidateClusterClassVariables validates clusterClassVariable.
+func ValidateClusterClassVariables(clusterClassVariables []clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, validateClusterClassVariableNamesUnique(clusterClassVariables, fldPath)...)
+
+	for i := range clusterClassVariables {
+		allErrs = append(allErrs, validateClusterClassVariable(&clusterClassVariables[i], fldPath.Index(i))...)
+	}
+
+	return allErrs
+}
+
+// validateClusterClassVariableNamesUnique validates that ClusterClass variable names are unique.
+func validateClusterClassVariableNamesUnique(clusterClassVariables []clusterv1.ClusterClassVariable, pathPrefix *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	variableNames := sets.NewString()
+	for i, clusterClassVariable := range clusterClassVariables {
+		if variableNames.Has(clusterClassVariable.Name) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					pathPrefix.Index(i).Child("name"),
+					clusterClassVariable.Name,
+					"variable names must be unique",
+				),
+			)
+		}
+		variableNames.Insert(clusterClassVariable.Name)
+	}
+
+	return allErrs
+}
+
+// validateClusterClassVariable validates a ClusterClassVariable.
+func validateClusterClassVariable(variable *clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Validate variable name.
+	allErrs = append(allErrs, validateClusterClassVariableName(variable.Name, fldPath.Child("name"))...)
+
+	// Validate schema.
+	allErrs = append(allErrs, validateSchema(&variable.Schema.OpenAPIV3Schema, variable.Name, fldPath.Child("schema", "openAPIV3Schema"))...)
+
+	return allErrs
+}
+
+// validateClusterClassVariableName validates a variable name.
+func validateClusterClassVariableName(variableName string, fldPath *field.Path) field.ErrorList {
```

Comment:
@killianmuldoon Can we please also block dots in variable names? 
xref: https://github.com/kubernetes-sigs/cluster-api/pull/5583#discussion_r749737758